### PR TITLE
[#3744] Use PQM to check PR approval and protected branch.

### PR DESCRIPTION
**Pull request by adiroiban**
Scope
=====

This update our merge_init build helper to no longer need changes-approved message as long as they are an explicit approve review message.

It also works with  the Github protected branched, to prevent a merge if the branch tests have not passed.

Changes
=======

Add support for the new review approve message API

Old messages are still supported.

Add support for using the GitHub branch protection.

How to try and test the changes
===============================

reviewers: @alibotean @lgheorghiu @brunogola 

Do the normal review process here.
When using the GitHub review comments you no longer need to add the 'changes-approved'  / 'needs-changes' marker.

`needs-review` is still required.

You can try various level of protections at https://github.com/chevah/brink/settings/branches/master

With the branch protection active for `status checks` you will need to always have a green `paver test_review` before doing a PQM ... as without the test_review the status is not updated.

The PQM is running the tests anyway... but it can't run the aggregated coverage tests. 

In the future I want to enforce 100% coverage for the tests code itself, to make sure there are no tests which are not executed.

Once this is merged, we need to release it and then include it in compat, server and other repors... and set up GitHub protected branches for then. chevah/compat already has the protected branch set up

### CODE REVIEW

**Comment by adiroiban** [comment]
needs-review

File: brink/qm.py [hunk 1]
```diff
@@ -248,12 +250,12 @@ def _getActionFromComment(comment):
     return 'no-action'
 
 
-def _get_environment(name, default=None):
+def _get_environment(name, default=_REQUIRED):
```
Comment (adiroiban) [inline brink/qm.py]: more here

**Comment by adiroiban** [comment]
needs-review

File: brink/qm.py [hunk 2]
```diff
@@ -116,11 +132,43 @@ def _review_properties(token, pull_id):
         repo = _get_repo(token=token)
         pull_request = _get_pull(repo, pull_id=pull_id)
 
+        branch_name = pull_request.head.ref
+        branch_sha = pull_request.head.sha.lower()
+
+        # Fail early if branch can not be merged.
+        if not pull_request.mergeable:
+            print("GitHub said that branch can not be merged.")
+            print("Please resole conflict and pull the changes.")
```
Comment (hcs0) [inline brink/qm.py]: resole -> resolve

File: brink/qm.py (hunk 1 above)
Comment (hcs0) [inline brink/qm.py]: less there

**Comment by hcs0** [comment]
Looks good.
Styleguide will need to be updated as well.

**Comment by hcs0** [comment]
I noticed this is under needs review on Trac.
Will it remain that way until the PR is merged?

**Comment by adiroiban** [comment]
needs-review

File: brink/qm.py [hunk 3]
```diff
@@ -370,11 +443,6 @@ def merge_commit(args):
         print("Review sha: %s" % (remote_sha,))
         sys.exit(1)
 
-    if not pull_request.mergeable:
-        print("GitHub said that branch can not be merged.")
-        print("Please resole conflict and pull the changes.")
```
Comment (hcs0) [inline brink/qm.py]: resole -> resolve

pull -> push the changes ?

File: brink/qm.py [hunk 4]
```diff
@@ -318,12 +397,6 @@ def merge_init():
         print("Review sha: %s %s" % (remote_sha, pr_branch_name))
         sys.exit(1)
 
-    # Fail early if branch can not be merged.
-    if not pull_request.mergeable:
-        print("GitHub said that branch can not be merged.")
-        print("Please resole conflict and pull the changes.")
```
Comment (hcs0) [inline brink/qm.py]: resole -> resolve

pull -> push the changes (?)

File: brink/qm.py [hunk 5]
```diff
@@ -116,11 +132,43 @@ def _review_properties(token, pull_id):
...
+            if not combined_status:
+                print('No branch status recorded by GitHub.')
+            else:
+                print('GitHub said that branch status is:')
+                for status in combined_status:
+                    print(" %s: %s (%s)" % (
+                        status.context, status.state, status.description))
+
+            print('The branch merge is blocked.')
+            print('Check GitHUB PR page for more details.')
```
Comment (brunogola) [inline brink/qm.py]: minor comment, but would be nice to use GitHub instead of GitH**UB** :)

File: brink/qm.py [hunk 6]
```diff
@@ -223,6 +278,26 @@ def _approvedByReviewer(reviewer, comments):
...
+        if state == u'COMMENTED':
+            # Just a comment. Can be ignored.
+            continue
+
+        if state == u'CHANGES_REQUESTED':
+            # Change requested before an approval.
+            # Not approved, even if the reviewer has approved it.
+            return False
+        if reviewer != author:
+            # Not a review from reviewer.
```
Comment (alibotean) [inline brink/qm.py]: This comment is out of sync withe the code. Either we change the condition or the comment itself.

Comment (adiroiban) [inline brink/qm.py]: the comment itself... `Hot a review from the targeted author`

**Comment by alibotean** [comment]
Just a minor comment, looking good.

**Pull request merged**